src/lib/device/device_manager.cpp

```diff
@@ -204,11 +205,12 @@ shared< VirtualDev > DeviceManager::create_vdev(vdev_parameters&& vparam) {
         pdevs.erase(pdevs.begin() + 1, pdevs.end()); // TODO: Pick random one
     }
 
-    // Adjust the maximum number chunks requested before round up vdev size.
-    uint32_t max_num_chunks = 0;
-    for (const auto& d : m_dev_infos) {
-        max_num_chunks += hs_super_blk::max_chunks_in_pdev(d);
-    }
+    // Based on the min chunk size, we calculate the max number of chunks that can be created in each target pdev
+    uint32_t min_chunk_size = hs_super_blk::min_chunk_size(vparam.dev_type);
+    // FIXME: it is possible that each vdev is less than max_num_chunks, but total is more than MAX_CHUNKS_IN_SYSTEM.
```

Comment:
This is a true statement, but it won't cause any problem, right?
`max_num_chunks` only garunteens chunk_size won't be smaller than the `min_chunk_size`. 
The actual `vparam.num_chunks` is a std::min of itself and this `max_num_chunks`, line: 223. 
And the input vparam.num_chunks is provided by HS consumer and this consumer can control all the services requested in HS won't exceed MAX_CHUNK_IN_SYSTEM, right?

Reply:
Yes it will not causing issue ,just fail the create vdev request.
I was thinking to tune the num_chunks based on the available chunk count. i.e if we have 1000 chunks available and the input requested 2000, we can tune that down to 1000... this is just a minor improvement .

Reply:
It should not fail the create vdev request, if the consumer of HS set the correct num_chunks for each services during hs.format_and_start, right? 

Your thought is also valid, that consumer just supply a total num_chunks consumer want to create and let HS do whatever trick to consume available chunks.

Reply:
Thinking about the case, we have 64K chunk limit across all vdevs. This 64K is a magic number that the client do not know. 
Even client knows like homeobject,  several service created with 
```
            HomeStore::instance()->format_and_start({
                {HS_SERVICE::META, hs_format_params{.dev_type = run_on_type, .size_pct = 5.0}},
                {HS_SERVICE::LOG, hs_format_params{.dev_type = run_on_type, .size_pct = 10.0, .chunk_size = 32 * Mi}},
                {HS_SERVICE::INDEX, hs_format_params{.dev_type = run_on_type, .size_pct = 5.0}},
                {HS_SERVICE::REPLICATION,
                 hs_format_params{.dev_type = run_on_type,
                                  .size_pct = 79.0,
                                  .num_chunks = 63000,
                                  .block_size = _data_block_size,
                                  .alloc_type = blk_allocator_type_t::append,
                                  .chunk_sel_type = chunk_selector_type_t::CUSTOM}},
            });
```

It is very hard to know if we are exceeding 64K,  as LOG device based on chunk_size, num_chunks is in-direct.  META/INDEX doesnt specified num_chunk but based on the default value of `hs_format_params`....


Later if LOG moved to dynamic size it is more tricky as the max size of the log device can be limited by the available chunk count left by other svc. e.g if data take 65500 then only 34 chunks left for log dev .

Reply:
I agree right now it is a magic number. We can get it through a static API to get the max_num_chunks from homestore (it can optimally to return the "real" max number, already taking min_chunk_size into account). 

I was thinking consumer need to specify num_chunks for all the services. In your case above, if we want to support only one service specify num_chunks, and let other services to land based on vdev size, then yes we also need to play some tricks in homestore.

Reply:
Yes, maybe lets ignore the well known number part for now, this is a easy fix issue whenever we want to fix.

>I was thinking consumer need to specify num_chunks for all the requested services. 
I feel very confused regarding the path  of dynamic size vdev introduced by LOG....  by dynamic sizing it naturally cannot provide num_chunks,  but we need to `reserve`  N  chunks for it to be dynamic within [0, N * CHUNK_SIZE)